### DISCUSSION TITLE
Valhalla Python bindings as a package

### DISCUSSION THREAD

**Discussion by nilsnolde** [valhalla/valhalla#3017] [Q&A]
I spent some time packaging valhalla for Python, which was ridiculously hard thanks to python's somewhat broken packaging infra.. The end result is platform-dependent packages which are distributed via binary `wheels` (and PyPI). I'd target the latest mac os, win64 and something called `manylinux_2_24` (basically compatible with all linuxes which have at least Debian 9 c/c++ base libs). The valhalla python package will also be able to build tiles, which it currently doesn't do, change the config during a session etc..

I wouldn't really want to PR the work (e.g. it's not optimally coded and in this particular case I'd like to be in full control of the functionality the bindings offer..). I don't see that as a problem though, as that wrapper code is super easy to maintain in a fork.

It'd still land on PyPI though. The only reasonable name is `valhalla(-router)` IMO..

Question is, is that ok for you? Or would you want us to obscure where the package is coming from so you don't end up getting issues about it by people getting confused?

P.S. I'm still debating with myself if I'd open the source or not (it _is_ a little ugly and I don't really want any PRs and maybe even some pay out of unsupported platforms). In any case, there'd be at least a PyPI registered repo where people can file issues.

**Comment by kevinkreiser** [comment]
to me it seems like the answer of the question will come from learning what you envision for the python artifacts long term. if you think one day you are going to open source it then i think its fine to name it something official sounding. if its going to remain closed source then maybe its best to name it in such a way as to make it obvious its not directly connected to this github org/repo. i hope that made sense :smile: 

fwiw i think its totally fine to PR something, get it merged but still have it marked as experimental in need of cleanup or as a proof of concept. these are totally ok states for things to be in when they are brand new. but of course the decision in that regard is completely up to you!

**Reply by nilsnolde** [reply]
Yeah, I'm leaning towards open source, most of the code is in a closed PR in this repo anyways (and ain't the best business model;)). So in that case it'd be okay for you to go with `valhalla`?

reg PR: thanks for the offer, but I think I also like to have it more obvious that we're pretty invested in valhalla (github has great SEO results;))

**Reply by kevinkreiser** [reply]
sure sounds good

**Comment by ImreSamu** [comment]
I don't know it will be also PyPy, GraalPython, Jython, IronPython compatible? 
(  there is a new Python C API proposal  HPy - https://hpyproject.org/  promising less complexity. https://hpyproject.org/blog/posts/2021/03/hello-hpy/  ) 
The PyPy support will be ideal. :smile:

**Reply by nilsnolde** [reply]
hahah nooo idea! never bothered with anything else other than cpython tbh.. pypy [might be possible](https://doc.pypy.org/en/latest/extending.html#cppyy) but I'd leave that to someone who wants to dig into it.

the others are c# or java based, so I guess not?!

**Comment by nilsnolde** [comment]
I finally had time to pretty much finalize this for now. ~~I opened the source here: https://github.com/gis-ops/valhalla-python~~ (closed for now)

I hardly did any changes to Valhalla core code other than the bindings. The only change I did do is a little sketchy (first file, graphreader.cc): https://github.com/gis-ops/valhalla-python/commit/32302611282d1223ef9da1e9b6fe030aa0422098. That was necessary to be able to switch tiles within a single python session. Not sure if that's a huge no-no?!

I already built all Python versions I'd like to support for Linux & Win in PyPI's test repo: https://test.pypi.org/project/valhalla/. OSX will come next week for whatever version Github Actions can build.

It's all in all a little awkward in terms of using the bindings, not very pythonic.. I can live with it, mostly planning to use the bindings in other downstream projects. Nevertheless I'd appreciate feedback in case anyone wants to try it before we go live with the packaging. @kevinkreiser and other owners: does that look ok to you from a licensing/naming etc perspective?

**Reply by escaped** [reply]
@nilsnolde Thanks for taking the time to create a proper python package. I used it already in one of my mini-project and it worked nicely :) Would love to use it in other projects as well, as it is much more convenient than running Valhalla in docker and making requests to the REST API. However, it seems like that the repo was either removed or made private, though the packages are still available on test-PyPI. Was wondering, whether you are still interested in publishing those packages to PyPi or whether the project was abandoned. :thinking:

**Reply by nilsnolde** [reply]
right, it's been closed for now. still not _entirely_ sure if I'll bring back the repository, I don't really want any issues, contributions etc. there's hardly anything different other than the setup.py, I contributed most of the other changes (waiting to be merged). honestly that's been _such_ a PITA to support all platforms.. we need it for a qgis-based project, that's why I created them. it'll definitely stay on pypi though, and likely move to the real one, or some other open repository, under a new name and very likely with a GPL license. a very similar osrm-py will join too:) that all will take another few months, it's an internal project so priority is unfortunately pretty low atm.
